Add guide for configuration am and ruler with mounted config maps

AkashSirimanna opened the pull request:
<!--  Thanks for sending a pull request!  Before submitting:

1. Read our CONTRIBUTING.md guide
2. Rebase your PR if it gets out of sync with master
3. Please do not edit/bump the Chart.yaml "version" field
-->

Context: The documentation on how to setup and run Alertmanager/Ruler is quite difficult to follow and not comprehensive. Many people have been running into issues regarding the sidecar. In fact, I was advised by some folks who contribute regularly to `cortex` to actually avoid using the sidecar. For this reason, I chose to add a guide on how to setup the ruler and alertmanager simply with configmaps which mount onto the container itself.

**What this PR does**: Adds a guide on how to configure cortex alertmanager and ruler using configmaps mounted onto the containers


**Checklist**
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

nschad commented:
Hey @AkashSirimanna,

Thank's a lot for the great guide. However for the ruler you could also have used this https://github.com/cortexproject/cortex-helm-chart/blob/master/templates/ruler/ruler-dep.yaml#L136 which is a bit easier than manually mounting every tenant configmap.

Helm Values on how to use that for the **ruler** (doesn't exists for alertmanager) exist here: https://cortexproject.github.io/cortex-helm-chart/guides/configure_rules_via_configmap.html

AkashSirimanna commented:
Hi @nschad thank you for the review.

I can add a section for this as well. Maybe the user can choose between the 2, depending on what works for them.

The reason I decided to go with the mounting approach was this: We would need to define the rules under the `ruler.directories.tenant` key. In order to do this, we would need to have the rules defined either in the helm-chart or in a separate file which is in the same repository as the helm chart. In our case, our rules were defined elsewhere in a different repo, this is I we chose to go with the mounting approach.

If you wanted to use the directory approach, we would need to have the rules files in the same repo as the helm chart, which is usually not the case. In the mounting approach, you are able to have them maintained elsewhere.

Nonetheless, I see the benefits of both sides, possibly I could add another section to the guide with how to setup the directories method?

nschad commented:
> ruler.directories.tenant

I actually would like to 

> Hi @nschad thank you for the review.
> 
> I can add a section for this as well. Maybe the user can choose between the 2, depending on what works for them.
> 
> The reason I decided to go with the mounting approach was this: We would need to define the rules under the `ruler.directories.tenant` key. In order to do this, we would need to have the rules defined either in the helm-chart or in a separate file which is in the same repository as the helm chart. In our case, our rules were defined elsewhere in a different repo, this is I we chose to go with the mounting approach.
> 
> If you wanted to use the directory approach, we would need to have the rules files in the same repo as the helm chart, which is usually not the case. In the mounting approach, you are able to have them maintained elsewhere.
> 
> Nonetheless, I see the benefits of both sides, possibly I could add another section to the guide with how to setup the directories method?

I actually would like to deprecate `ruler.directories.tenant` because the alertmanager doesn't have it and I don't think its necessary thing to have

AkashSirimanna commented:
I see. If thats the case, what do you think of this guide then? If theres anything to add/remove/modify let me know.

nschad commented on docs/guides/configure_ruler_and_alertmanager_with_configmap.markdown:
```diff
@@ -0,0 +1,179 @@
+        name: tenantA-alertmanager-config
+      name: tenantA-alertmanager-config
+  extraVolumeMounts:
+    - name: tmp
+      mountPath: /tmp
```
Same as above. We should really add that to the chart as the default. I keep forgetting it :(

nschad commented on docs/guides/configure_ruler_and_alertmanager_with_configmap.markdown:
```diff
@@ -0,0 +1,179 @@
+
+### Configuraion For Ruler and Alertmanager
+Add this to the `values.yaml`.
+Note that when we mount `tenantA-ruler-config`, we mount it under a folder which should be named after the tenant it corresponds to. In this situation, since we want `tenantA-ruler-config` to be the alerting rules for `tenantA`, we mount it under `/tmp/rules/tenantA`.
```
```suggestion
Note that when we mount `tenantA-ruler-config`, we mount it under a folder which should be named after the tenant it corresponds to. In this situation, since we want `tenantA-ruler-config` to be the alerting rules for `tenantA`, we mount it under `/rules/tenantA`.
```

nschad reviewed (commented):
Great Guide. I just had some few points

AkashSirimanna commented on docs/guides/configure_ruler_and_alertmanager_with_configmap.markdown:
```diff
@@ -0,0 +1,179 @@
+  ruler:
+    enable_api: true
+    enable_alertmanager_discovery: true
+    alertmanager_url: "http://_http-metrics._tcp.cortex-alertmanager/api/prom/alertmanager/"
```
Nope, removed 🙏

AkashSirimanna commented:
@nschad Thank you for the suggestions, I incorporated them and also changed `tenantA` to `tenantfoo`, i realized that capital letters in config map names are invalid.

I'll squash the commits once the changes look good

AkashSirimanna commented:
Good to go 😄

nschad commented on docs/guides/configure_ruler_and_alertmanager_with_configmap.markdown:
```diff
@@ -0,0 +1,179 @@
+      mountPath: /tmp/rules/tenantA
+alertmanager:
+  extraVolumes:
+    - name: tmp
+      emptyDir: {}
```
@AkashSirimanna What about this?

AkashSirimanna replied:
Oops, forgot about this, yea I think you're right. I'm pretty sure it just needs a intermediate folder that is read-write

nschad reviewed (commented):
Looks good however I won't get to testing it today but probably tomorrow :)

AkashSirimanna commented:
> Looks good however I won't get to testing it today but probably tomorrow :)

Sounds good! Im going to test it out in a bit to make sure everything is good

AkashSirimanna commented:
Tested on a local minikube cluster and it seems to be working :) I'll check back once you've got a chance to test it :)

nschad commented on docs/guides/configure_ruler_and_alertmanager_with_configmap.markdown:
```diff
@@ -0,0 +1,171 @@
+    - configMap:
+        defaultMode: 420
+        name: tenantfoo-alertmanager-config
+      name: tenantfoo-alertmanager-config
```
```suggestion
      name: alertmanager-config
```

needs to match what is in extraVolumeMounts

nschad reviewed (changes requested):
Guide verified! LGTM! However found a tiny typo

nschad reviewed (approved):
@AkashSirimanna Cool. I fixxed the Git mess I caused. Sorry about that. You probably wanna do
```
git fetch origin am-ruler-configmap-guide
git reset --hard origin/am-ruler-configmap-guide
```

but you are done so probably no need :)

AkashSirimanna commented:
> @AkashSirimanna Cool. I fixxed the Git mess I caused. Sorry about that. You probably wanna do
> 
> ```
> git fetch origin am-ruler-configmap-guide
> git reset --hard origin/am-ruler-configmap-guide
> ```
> 
> but you are done so probably no need :)

Ah no worries, I probably helped mess it up 🤣 .
Thanks for the review, hopefully this helps the community out a bit.

kd7lxl commented on docs/guides/configure_ruler_and_alertmanager_with_configmap.markdown:
```diff
@@ -0,0 +1,172 @@
+
+# Setup
+
+In order to get this working, we must do a couple things
```
```suggestion
In order to get this working, we must do a couple things.
```

kd7lxl commented on docs/guides/configure_ruler_and_alertmanager_with_configmap.markdown:
```diff
@@ -0,0 +1,172 @@
+
+To provide the Ruler with the *rules* to alert on, we must create a ConfigMap with the desired ruleset. How you maintain and deploy the ConfigMap is completely in your control, the only requirement is that the ConfigMap be on the same namespace as the Ruler.
+
+**IMPORTANT** things to note
```
```suggestion
**IMPORTANT** things to note:
```

kd7lxl commented on docs/guides/configure_ruler_and_alertmanager_with_configmap.markdown:
```diff
@@ -0,0 +1,172 @@
+**IMPORTANT** things to note
+
+- This ruleset is ONLY for tenantfoo (We will explicitly map it to tenantfoo later)
+- The section under `tenantfoo.yaml` is validated and interpreted exactly as a Prometheus rules configuration would be. How you configure this is completely up to you.
```
Let's include a reference to the documentation here.
```suggestion
- The section under `tenantfoo.yaml` is validated and interpreted exactly as a [Prometheus rules configuration](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/) would be. How you configure this is completely up to you.
```

nschad commented:
@AkashSirimanna Okay LOL.

I keep finding Issues but I think I got them all now (some of them have nothing to-do with you).

kd7lxl commented on docs/guides/configure_ruler_and_alertmanager_with_configmap.markdown:
```diff
@@ -0,0 +1,169 @@
+
+To provide the Alertmanager with the information needed to *route* notifications to stakeholders, we must create a ConfigMap with the desired routing rules. How you maintain and deploy the ConfigMap is completely in your control, the only requirement is that the ConfigMap be on the same namespace as the Alertmanager.
+
+**IMPORTANT** things to note
```
```suggestion
**IMPORTANT** things to note:
```

kd7lxl commented on docs/guides/configure_ruler_and_alertmanager_with_configmap.markdown:
```diff
@@ -0,0 +1,169 @@
+**IMPORTANT** things to note
+
+- These routing rules are ONLY for tenantfoo. The Alertmanager takes the key under `data` (excluding `.yaml`) and uses it as the tenant name for that specific config.
+- The section under `tenantfoo.yaml` is validated and interpreted exactly as a Prometheus alerting rules configuration would be. How you configure this is completely up to you.
```
```suggestion
- The section under `tenantfoo.yaml` is validated and interpreted exactly as a [Prometheus alerting rules configuration](https://prometheus.io/docs/prometheus/latest/configuration/alerting_rules/) would be. How you configure this is completely up to you.
```

nschad commented on docs/assets/scripts/copyCode.js:
```diff
@@ -6,7 +6,7 @@ const template = `
 
 document.addEventListener("DOMContentLoaded", function () {
-  const codeBlocks = document.querySelectorAll('.highlighter-rouge');
+  const codeBlocks = document.querySelectorAll('div.highlighter-rouge');
```
about time, tehe

AkashSirimanna commented:
Thank you @kd7lxl, ill go ahead and make the changes
